Adding a viral read identification subworkflow for ONT data.

simonleandergrimm opened the pull request:
**Subworkflow**
 - Created new EXTRACT_VIRAL_READS_ONT workflow for ONT data processing
 - Renamed existing EXTRACT_VIRAL_READS to EXTRACT_VIRAL_READS_SHORT for clarity
**Processes** 
 - Added DUSTMASKER_FASTQ_GZIPPED process for low-complexity filtering of ONT reads
 - Added MERGE_SAM process to combine minimap2 sam files.
 - Added PROCESS_VIRAL_MINIMAP2_SAM process to extract viral read information

**Scripts**
 - Wrote a new Python script ([process_viral_minimap2_sam.py](https://github.com/naobservatory/mgs-workflow/compare/simon-minimap-indices...simon-ont-extract-viral?expand=1#diff-f99076ccd4d94a164c962c5456bdc7255ca6e044fe243abe70c0159007424cb9)), used in PROCESS_VIRAL_MINIMAP2_SAM, which gives an output TSV containing minimap2-generated HV alignments, filtered to the relevant host taxon.

**Containers**
    - Added new container definitions (minimap2_samtools and samtools)
    - Added new container for pysam.

simonleandergrimm commented:
@willbradshaw I haven't yet added tests to this branch, as I first want to get your high-level take on the approach I'm taking here. Can you do an initial review and let me know what you think? I'll do some further work on the branch today but I don't expect to make major changes prior to your review. 

I'll post some results of the pipeline here too.

willbradshaw commented on modules/local/processViralMinimap2Sam/main.nf:
```diff
@@ -0,0 +1,28 @@
+        '''
+        hv_sam=!{hv_sam}
+        clean_reads=!{clean_matched_subset_merged_ch}
+        gunzip -c ${clean_reads} > ${clean_reads}.fastq
```
Also I think the current code will get you an input suffix of `.fastq.gz.fastq`, which might trip up some functions.

simonleandergrimm replied:
Fixed.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+                meta_db["genome_id"],
+                meta_db["taxid"],
+                meta_db["species_taxid"],
+                meta_db["organism_name"]
```
Why add this here?

simonleandergrimm replied:
can you say more?

simonleandergrimm replied:
Main reason for including this is that I find it useful to have the organism names in the final HV TSV.

willbradshaw replied:
But you can easily get it from the viral taxonomy database, which you'll want to be joining with this database in most analysis situations anyway, in order to access taxonomy information.

In general, the barrier for changing the format of the hits DB compared to the paired version should be quite high, and I don't think this meets it.

simonleandergrimm replied:
```
format of the hits DB
```
You mean the output of extractViralReads{Short/ONT}?

I find this info pretty useful (as you say, I use it in most analysis situations). Is this something where we'd want to have this also happen in the paired version?

willbradshaw replied:
> You mean the output of extractViralReads{Short/ONT}?

The viral read TSV, yes.

My objection to this is twofold:

1. It's redundant with information in the viral taxonomy DB, which is readily available and easy to join with this output for downstream analysis.
2. Worse, I'm not confident it will always use the same name as the taxonomy DB (which is the source of truth here) for every taxid, which could lead to nasty mismatches and confusion.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+        virus_taxa = set(virus_db["taxid"].values)
+        print_log(f"Virus DB imported. {len(virus_db)} total viral taxids.")
+
+        virus_status_dict = {
```
Flagging for discussion (as something that's not in the original Bowtie2 script)

simonleandergrimm replied:
Do you want discus this in person? I'll be remote Monday but happy to chat.

willbradshaw replied:
I made another comment up where the dict is actually used, so this one can be resolved

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+    with open(out_file, "w") as out_fh:
+        header = None
+        try:
+            with pysam.AlignmentFile(sam_file, "r") as sam_file:
```
This doesn't work for the Bowtie2 script because we're writing SAM without the headers (which is necessary because otherwise the SAM file gets way too big). I expect you should probably be doing the same for Minimap2. Have you checked to see how this affects the size of the SAM file?

simonleandergrimm replied:
Checking the ouput of the virus and human minimap2 SAM output, I get files (gzipped) that are between 300 and 500KB. This is small enough to not matter much? I find working with pysam a lot cleaner compared to manual parsing.

```
[ec2-user@ip-172-31-40-150 work]$ ls -lh cf/833402a43cca87afe673a458091f9b/ | grep hv_minimap2_mapped.sam
-rw-r--r--. 1 root     root     326K Mar  3 15:17 NAO-ONT-20250213-Zephyr10-05-div0040_hv_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ ls -lh 20/f809a33134844f6f546000577b2d3d/ | grep hv_minimap2_mapped
.sam
-rw-r--r--. 1 root     root     328K Mar  3 15:17 NAO-ONT-20250213-Zephyr10-05-div0026_hv_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ ls -lh ab/840fbb3c6a16f05b8d8bf02be7c705/ | grep hv_minimap2_mapped
.sam
-rw-r--r--. 1 root     root     336K Mar  3 15:17 NAO-ONT-20250213-Zephyr10-05-div0025_hv_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ ls -lh 3d/a621a0c584c0a7c014c503b27ab757/ | grep hv_minimap2_mapped
.sam
[ec2-user@ip-172-31-40-150 work]$ ls -lh 3d/a621a0c584c0a7c014c503b27ab757/ | grep human_minimap2_map
ped.sam
-rw-r--r--. 1 root     root     198K Mar  3 15:16 NAO-ONT-20250213-Zephyr10-05-div0040_human_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ ls -lh fc/7446d97991f949339ee5d6632a4a3e/ | grep human_minimap2_map
ped.sam
-rw-r--r--. 1 root     root     420K Mar  3 15:16 NAO-ONT-20250213-Zephyr10-05-div0026_human_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ ls -lh 6c/8aafe8f910d5c568c6a9ff6fdb89ac/ | grep human_minimap2_map
ped.sam
-rw-r--r--. 1 root     root     503K Mar  3 15:16 NAO-ONT-20250213-Zephyr10-05-div0025_human_minimap2_mapped.sam.gz
[ec2-user@ip-172-31-40-150 work]$ 
```

willbradshaw replied:
I agree those file sizes are not concerning. How big was your input? E.g. does this represent a MinION run, a PromethION run, or something else?

I also still think it would be good to check what fraction of those files is made up of header lines. I expect the number of header lines to increase quite quickly as the amount of data increases.

simonleandergrimm replied:
This is a Promethion run. The input files are fairly small, but this is generally true of Nanopore: the total read numbers are lower, and given how we are doing basecalling, each barcode has several corresponding FASTQs.

I just ran the pipeline on two larger FASTQs from a Promethion PHC run. The input file was 159 MB. The output HV SAM is 352K, the output Human SAM is 463K. 

The human SAM headers shouldn't get larger (their reference genomes are just human chromosomes). HV should also stay small. With ribosomal SAMs this could change, but this is not what we are working with here.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+    )
+    return parser.parse_args()
+
+def main():
```
A general comment: where this script deviates from the original bowtie2 script, you should have comments explaining why.

simonleandergrimm replied:
Will add.

willbradshaw replied:
I meant comments in the script btw, not on Github

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+                    if read.is_unmapped or read.is_secondary or read.is_supplementary:
+                        continue
+                    read_id = read.query_name
+                    read_seq = clean_read_dict[read_id]
```
Is this necessary? The Bowtie2 SAM just includes the read sequence.

willbradshaw replied:
OK, that makes sense, but you should also include the read sequence the aligner was actually run on alongside it.

simonleandergrimm replied:
Fixed.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+                        continue
+                    if header is None:
+                        out_fh.write(join_line(line.keys()))
+                        header = True
```
I don't understand what you're doing with `header` in this function. Can you explain?

simonleandergrimm replied:
For the first iteration of this for loop, `header` is set to None. It recognizes this and sets the column headers to the keys in the line dictionary (the keys are set in "parse_sam_alignment"). `header` is then set to True, and this condition is not triggered again. So it's really just a Boolean, nothing else.

willbradshaw replied:
OK, I see. See my comment below, I think you can skip all this logic by just writing the header line before you start the for loop.

simonleandergrimm replied:
Done.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+        out["query_sequence_clean"] = clean_sequence
+
+    out["sample_name"] = read.get_tag("RG")
+    out["minimap2_genome_id_primary"] = reference_genome_name
```
"primary" as opposed to what?

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+    out["minimap2_ref_end"] = read.reference_end
+    out["minimap2_alignment_start"] = read.query_alignment_start
+    out["minimap2_alignment_end"] = read.query_alignment_end
+    out["minimap2_cigar"] = read.cigarstring
```
There are various fields from the Bowtie2 script that are missing here (edit distance, map quality, etc); are these not in the Minimap2 SAM?

simonleandergrimm replied:
`minimap2_primary_alignment_score` is actually map quality, will change. I also added edit distance.

simonleandergrimm replied:
Done

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,213 @@
+    out["minimap2_taxid_primary"] = reference_taxid
+
+    out["minimap2_read_length"] = read.query_length
+    out["minimap2_primary_alignment_score"] = read.mapping_quality
```
If you're planning on doing any length normalization for the alignment score for downstream filtering (which I'm guessing you probably should) you should include it in this script somewhere.

simonleandergrimm replied:
The current plan is to do downstream filtering with BLAST, so this is probably not needed here?

willbradshaw commented on modules/local/pulloutFastq/main.nf:
```diff
@@ -0,0 +1,23 @@
+// Pull out original clean reads from cleaned fastq, based on smaller FASTQ read ids
```
Is this because you want to unmask the reads?

simonleandergrimm replied:
Yes

willbradshaw replied:
The process looks fine, but the name is too vague and the argument names are too specific. This is a general process for extracting reads from one FASTQ that share an ID with a read in another fastq.

Consider calling it something like `EXTRACT_SHARED_FASTQ_READS` and renaming the arguments to `fastq_ids` and `fastq_seqs`, or even `fastq_1` and `fastq_2`.

willbradshaw commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+        // Generate HV TSV
+        merged_virus_sam_ch = MERGE_SAM_VIRUS(virus_sam_ch, "hv")
+        hv_tsv_ch = PROCESS_VIRAL_MINIMAP2_SAM(merged_virus_sam_ch, clean_matched_subset_merged_ch, genome_meta_path, virus_db_path, host_taxon)
+
```
Why did you decide to drop the filtering step?

simonleandergrimm replied:
I want to use BLAST for filtering, given the low number of viral reads. So I didn't add a different filtering step.

simonleandergrimm commented:
Thanks for the comments. Some of the edits I've made over the past week were quite WIP, should have flagged in the meantime... Will get back to your comments shortly.

simonleandergrimm commented on modules/local/filtlong/main.nf:
```diff
@@ -3,13 +3,17 @@ process FILTLONG {
         tuple val(sample), path("${sample}_filtlong.fastq.gz"), emit: reads
     shell:
-        // Filter reads based on length (min 100 bp) and mean average base quality (min 99%, i.e, a Phred score of 20)
+        // Filter reads based on min/max length and min mean quality
```
This was due to very long (erroneous) reads in fastq files containing non-demuxed reads, which broke fastqc.

simonleandergrimm commented on subworkflows/local/extractViralReadsShort/main.nf:
As mentioned at the start of the PR, I first wanted to check if the overall approach makes sense before writing tests.

willbradshaw replied:
Sure, but there's a difference between writing new tests and checking old ones aren't broken.

simonleandergrimm replied:
Fixed test!

simonleandergrimm commented on workflows/run_dev_se.nf:
```diff
@@ -23,29 +27,43 @@ nextflow.preview.output = true
-    PROFILE(SUBSET_TRIM.out.trimmed_subset_reads, kraken_db_path, params.ref_dir, "0.4", "27", "ribo",
-        params.bracken_threshold, params.single_end)
+    if ( params.ont ) {
+        bracken_ch = Channel.empty()
```
Because I haven't evaluated how well profiling works yet on ONT.

simonleandergrimm commented on workflows/run_dev_se.nf:
```diff
@@ -67,12 +85,16 @@ workflow RUN_DEV_SE {
-        RUN_QC.out.qc_qbase >> "results"
-        RUN_QC.out.qc_qseqs >> "results"
-        RUN_QC.out.qc_lengths >> "results"
+        // RUN_QC.out.qc_basic >> "results"
```
This was a WIP change.

simonleandergrimm replied:
Fixed

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+        blast_db_path = "${params.ref_dir}/results/${params.blast_db_prefix}"
+
+        // Filter reads by length and quality scores
+        filtered_ch = FILTLONG(reads_ch, 50, 10000, 90)
```
No reason, I turned off SUBSET_TRIM and thus haven't edited it yet. They should be the same.

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+        no_human_ch = human_minimap2_ch.reads_unmapped
+
+        // Identify other contaminants
+            // contam_minimap2_ch = MINIMAP2_CONTAM(no_human_ch, minimap2_contam_index, "other", false)
```
Because I hadn't yet created the contam minimap2 index. Did now.

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+            // no_contam_ch = contam_minimap2_ch.reads_unmapped
+
+        // Identify virus reads
+        virus_minimap2_ch = MINIMAP2_VIRUS(no_human_ch, minimap2_virus_index, "hv", false)
```
So I can get the SAM output from MINIMAP2_VIRUS, using the FASTQ output through contaminant removal would remove the mapping information. I could get it back but that seems somewhat complicated. Let me know if I'm overlooking downsides here.

simonleandergrimm replied:
But it could make sense as the HV read set is probably a lot smaller, making the filtering step faster. Still as noted in the comment below, the ONT files are small enough that the difference shouldn't be very large.

willbradshaw replied:
Ah, that makes sense. In the short read version the filtered FASTQs go into Kraken and then get merged back into the processed SAM file later, but that doesn't apply in this case. I think that's a sufficient justification.

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+
+        // Generate HV TSV
+        merged_virus_sam_ch = MERGE_SAM_VIRUS(virus_sam_ch, "hv")
+        hv_tsv_ch = PROCESS_VIRAL_MINIMAP2_SAM(merged_virus_sam_ch, clean_matched_subset_merged_ch, genome_meta_path, virus_db_path, host_taxon)
```
Fair point, but currently, an average PHC run gives like 3000-4000 virus matches, Zephyr 100-7000, so the efficiency gain wouldn't be very high.

willbradshaw replied:
It still seems to be like tech debt we don't need to be incurring.

simonleandergrimm replied:
Yeah, that seems right. Will address.

simonleandergrimm replied:
Done

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+        // Mark non-complex read sections
+        masked_ch = DUSTMASKER_FASTQ_GZIPPED(filtered_ch)
+
+        // Drop human reads before pathogen identification
```
Not very, but runtime is very fast. I ran ~7 Zephyr datasets through the pipeline today. A dataset that's about 1/4 the size of a large PHC ONT dataset takes 39m:
```
Completed at: 27-Feb-2025 22:33:28
Duration    : 39m 13s
CPU hours   : 33.6
Succeeded   : 1'383
```

willbradshaw replied:
39m is long enough that I think it would be worth experimenting. (Though the very small number of CPU-hours makes me think the main bottlenecks might be slow single-threaded scripts.)

simonleandergrimm replied:
Just so I understand `slow single-threaded scripts` would this be things like Python scripts?

simonleandergrimm commented on modules/local/minimap2/main.nf:
```diff
@@ -49,7 +50,7 @@ process MINIMAP2 {
         #   - Second branch (samtools view -u -F 4 -) filters SAM to aligned reads and saves FASTQ
         #   - Third branch (samtools view -h -F 4 -) also filters SAM to aligned reads and saves SAM
-        zcat !{reads} \
+        zcat ${reads} \
```
Fixed.

simonleandergrimm commented on modules/local/processViralMinimap2Sam/main.nf:
```diff
@@ -0,0 +1,28 @@
+    label "single"
+    input:
+        path hv_sam
+        path clean_matched_subset_merged_ch
```
Fixed.

simonleandergrimm commented on workflows/run_dev_se.nf:
```diff
@@ -10,11 +10,15 @@ import java.time.LocalDateTime
 include { SUBSET_TRIM } from "../subworkflows/local/subsetTrim"
 include { RUN_QC } from "../subworkflows/local/runQc"
 include { PROFILE } from "../subworkflows/local/profile"
+if ( params.ont ) {
```
This creates problems for the output, as either process won't have anything to output. Unsure if the added complexity of adding empty Channels and having two output statements for each workflow is worth it?

simonleandergrimm replied:
Hm, right. Fixed!

willbradshaw replied:
Can you now remove the if statement in the imports?

simonleandergrimm replied:
Done

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,65 @@
+        filtered_ch = FILTLONG(reads_ch, 50, 10000, 90)
+
+        // Mark non-complex read sections
+        masked_ch = DUSTMASKER_FASTQ_GZIPPED(filtered_ch)
```
Started switching to bbmask.sh

simonleandergrimm commented on modules/local/processViralMinimap2Sam/main.nf:
```diff
@@ -0,0 +1,28 @@
+        host=!{host_taxon}
+        process_viral_minimap2_sam.py -s ${hv_sam} -r ${clean_reads}.fastq -m ${metadata} -v ${virus_db} -ht ${host} -o ${out}
+        # Link input to output for testing
+        ln -s !{hv_sam} minimap2_sam_filtered.sam
```
Done

simonleandergrimm commented on modules/local/dustmasker/main.nf:
Done

simonleandergrimm commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+
+# Alignment-level functions
+
+def parse_sam_alignment(read, genbank_metadata, viral_taxids, virus_status_dict, clean_sequence, sample):
```
This is using pysam instead of more manual value extraction.

simonleandergrimm commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+    return line
+
+
+def extract_viral_taxid_and_name(genome_id, gid_taxid_name_dict, viral_taxids):
```
Similar to extract_viral_taxid, but also adding name of the taxonomic assignment (e.g., Human group 1 coronavirus)

willbradshaw replied:
See below, I don't think this (the name) is worth including

simonleandergrimm replied:
Will drop the name.

simonleandergrimm commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+        required=True,
+        help="Path to Minimap2 SAM file."
+    )
+    parser.add_argument(
```
FASTQ file that contains the non-masked version of the HV reads.

simonleandergrimm replied:
Done

simonleandergrimm commented:
@willbradshaw This is ready for re-review. The biggest question is if there are major things that need to change. If not, I will both start refining + add tests. I will talk with Jeff today about adding in BLAST.

willbradshaw commented:
> I will talk with Jeff today about adding in BLAST

Note, you need to talk to me about this, not just Jeff.

willbradshaw commented on modules/local/pulloutFastq/main.nf:
```diff
@@ -0,0 +1,23 @@
+        set -euo pipefail
+
+        # Get mapped read ids
+        zcat !{mapped_fastq} | awk 'NR%4==1' | sed 's/^@//' > mapped_read_ids.txt
```
An alternative here is `seqtk comp`

`seqtk comp !{mapped_fastq} | cut -f 1 > read_ids.txt`

simonleandergrimm replied:
Great, replaced.

willbradshaw commented on modules/local/pulloutFastq/main.nf:
```diff
@@ -0,0 +1,23 @@
+// Pull out original clean reads from cleaned fastq, based on smaller FASTQ read ids
+process PULLOUT_FASTQ {
```
Note somewhere that this only works with gzipped fastqs

simonleandergrimm replied:
Done

willbradshaw commented on modules/local/pulloutFastq/main.nf:
```diff
@@ -0,0 +1,23 @@
+
+        # Link input to output for testing
+        ln -s !{mapped_fastq} !{sample}_mapped_in.fastq.gz
+        ln -s !{cleaned_fastq} !{sample}_cleaned_in.fastq.gz        '''
```
Terminal ''' should be on a new line

simonleandergrimm replied:
Fixed

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+        default=sys.stdin,
+        help="Path to clean reads (default: stdin)."
+    )
+    parser.add_argument(
```
What is this for? It's not in the Bowtie2 script.

simonleandergrimm replied:
I actually don't need this anymore, thanks for spotting.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+                    out_fh.write(join_line(line.values()))
+
+                if header is None:
+                    out_fh.write("query_name\tminimap2_name_primary\tquery_sequence_clean\tsample_name\tminimap2_genome_id_primary\tminimap2_taxid_primary\tminimap2_read_length\tminimap2_map_qual\tminimap2_ref_start\tminimap2_ref_end\tminimap2_alignment_start\tminimap2_alignment_end\tminimap2_cigar\tminimap2_edit_distance\n")
```
I would write these before you start the for loop at all.
If you want to avoid errors, you could try defining a vector of header strings, join and write those, then assert that the keys for each line match the headers.

simonleandergrimm replied:
Good idea, added.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+        # Process SAM
+        print_log("Processing SAM file...")
+        process_sam(sam_file, out_file, sample, gid_taxid_name_dict, virus_taxa, virus_status_dict, clean_read_dict)
+        print_log("Zipping output file...")
```
I don't think this is doing anything

simonleandergrimm replied:
Dropped.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+        description="Process Minimap2 SAM output into a TSV with viral alignment information."
+    )
+    parser.add_argument(
+        "-sa", "--sam",
```
single-hyphen arguments should be single characters (they can get chained together)

simonleandergrimm replied:
Fixed

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+    )
+    parser.add_argument(
+        "-r", "--reads",
+        type=lambda f: open_by_suffix(f, "r"),
```
If you're keeping the pattern of opening files in the argparse block, you need a `finally` block at the end of main() to make sure they get closed.

simonleandergrimm replied:
Done

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+    with open_by_suffix(out_file, "w") as out_fh:
+        header = None
+        try:
+            with pysam.AlignmentFile(sam_file, "r") as sam_file:
```
Can this handle compressed SAM files?

simonleandergrimm replied:
Code:
```
#!/usr/bin/env python3
import pysam

sam_file = "/home/ec2-user/code/mgs-workflow/test-alignments.sam.gz"

with pysam.AlignmentFile(sam_file, "r") as sam_file:
    for read in sam_file:
        print(read.query_name)
```

Output:
```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ ./test-sam-reading.py 
Header contains 69073 reference sequences
3af1e54c-9e88-4a36-a0d8-1963fb30ca87
9fc87db5-2a7b-4b26-b0fb-7e8cc0e887d1
cfa80705-3f49-44db-9614-551b6a7f0d73
a268bfa1-6e0c-414e-b31c-8d3eda35c4f3
b67fa54e-3063-41ed-973e-214928f7a14c
b67fa54e-3063-41ed-973e-214928f7a14c
b67fa54e-3063-41ed-973e-214928f7a14c
[ec2-user@ip-172-31-40-150 mgs-workflow]$ 
```

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+    line["minimap2_name_primary"] = reference_name
+    if read.is_reverse:
+        # When minimap2 maps to the RC version of a strand, if returns the RC version of the read
+        line["query_sequence_clean"] = clean_sequence.reverse_complement()
```
Suggest `query_seq_clean` to match the Bowtie2 output.

Also you should add the quality scores and lengths (`query_qual_...` and `query_len_...`).

simonleandergrimm replied:
Added

willbradshaw replied:
Please change to `query_seq_clean` and `query_qual_clean`.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,230 @@
+    line["minimap2_alignment_end"] = read.query_alignment_end
+    line["minimap2_cigar"] = read.cigarstring
+    line["minimap2_edit_distance"] = read.get_tag("NM")
+
```
I note the lack of any kind of alignment score metric here. This seems like a pretty important thing to record?

simonleandergrimm replied:
Added.

simonleandergrimm replied:
Added

simonleandergrimm commented on configs/containers.config:
```diff
@@ -130,7 +140,23 @@ process {
+        //- bioconda
+        //dependencies:
+        //- pip
+        //- pip:
```
The formatting was somewhat wrong. Here's what seqera containers gives me
```
channels:
- conda-forge
- bioconda
dependencies:
- pip
- pip:
  - biopython==1.85
  - pandas==2.2.3
  - pysam==0.23.0
```

simonleandergrimm commented on configs/containers.config:
```diff
@@ -130,7 +140,23 @@ process {
     withLabel: minimap2 {
         container = "staphb/minimap2:2.28"
     }
+    withLabel: samtools {
```
Originally used in MERGE_SAM, which we don't need anymore.

simonleandergrimm commented on modules/local/concatenateFastq/main.nf:
Deleted

simonleandergrimm commented on modules/local/dustmasker/main.nf:
Done

simonleandergrimm commented on modules/local/samtools/main.nf:
Removed!

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,72 @@
+include { PROCESS_VIRAL_MINIMAP2_SAM } from "../../../modules/local/processViralMinimap2Sam"
+include { BLAST_VIRAL } from "../../../subworkflows/local/blastViral"
+include { PULLOUT_FASTQ } from "../../../modules/local/pulloutFastq"
+include { CONCATENATE_FASTQ_GZIPPED } from "../../../modules/local/concatenateFastq"
```
Removed.

simonleandergrimm commented on subworkflows/local/makeVirusGenomeDB/main.nf:
```diff
@@ -7,7 +7,7 @@ include { FILTER_VIRAL_GENBANK_METADATA } from "../../../modules/local/filterVir
 include { CONCATENATE_GENOME_FASTA } from "../../../modules/local/concatenateGenomeFasta"
 include { FILTER_GENOME_FASTA } from "../../../modules/local/filterGenomeFasta"
-include { MASK_GENOME_FASTA } from "../../../modules/local/maskGenomeFasta"
+include { MASK_GENOME_FASTA } from "../../../modules/local/maskRead"
```
Went back to original name.

simonleandergrimm commented on modules/local/maskRead/main.nf:
```diff
@@ -30,6 +30,35 @@ process MASK_GENOME_FASTA {
+    input:
+        tuple val(sample), path(reads)
+        val(window_size)
+	    val(entropy)
```
Fixed

simonleandergrimm commented on modules/local/maskRead/main.nf:
```diff
@@ -30,6 +30,35 @@ process MASK_GENOME_FASTA {
 	'''
 }
+
+// Mask low complexity FASTQ read regions
```
Separated the two processes into separate files.

simonleandergrimm commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,246 @@
+        clean_query_record = clean_query_record.reverse_complement()
+
+    query_seq_clean = clean_query_record.seq
+    query_qual_clean = clean_query_record.letter_annotations["phred_quality"]
```
Still need to fix this so it returns symbols, not numbers.

simonleandergrimm replied:
Done

simonleandergrimm commented:
> In terms of major design decisions, the big one right now is how to handle filtering and, relatedly, what kind of metrics to return to the user so they can make downstream filtering decisions themselves.

I've gotten back to your comments. I'll next check if filtering can be done with minimap2 output alone; most likely we'll need BLAST anyway for human read filtering.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,246 @@
+    reference_taxid, reference_name = extract_viral_taxid_and_name(reference_genome_name, genbank_metadata, viral_taxids)
+
+    # Filtering out non-host-taxon reads
+    if virus_status_dict[reference_taxid] == "0":
```
Ah, I see. This seems to be the only place in the script where `virus_status_dict` is actually used.

I wonder if this is actually necessary? There's no equivalent behavior in the Bowtie2 script, so I'm worried about the possibility of confusion if it's only here. And it's easy to do this filtering downstream using the virus taxonomy DB.

simonleandergrimm replied:
Hm, can we maybe drop the filtering but keep adding the host_status? I believe this is what we do in extractViralReadsShort.

simonleandergrimm replied:
Agreed, but this gets added in process_kraken_viral.py.

simonleandergrimm replied:
Hm ok. I'd be somewhat pro leaving this in, but let me know what your preference is.

willbradshaw replied:
If you're not filtering on it (which I think you shouldn't) we should drop it. As with the name, you can get it from the virus taxonomy DB, and we shouldn't be having redundant information in multiple places.

willbradshaw commented:
@simonleandergrimm I spoke to Jeff about filtering.

- We don't need to add a new BLAST process (you can just forward the output of this to the existing BLAST_VIRAL workflow) 
- We can leave off any initial k-mer based filtering for now
- For alignment-score-based filtering, I still think this is something you should look into, but we can skip it for now in the interest of getting this merged. Can you make an issue for it? Once you've done that I'll resolve the relevant conversations.

simonleandergrimm commented:
@willbradshaw Created an issue here: https://github.com/naobservatory/mgs-workflow/issues/224

Also added BLAST. I will clean up the branch and add tests tomorrow.

simonleandergrimm commented:
Alright, I've now added tests for the new processes and subworkflow. This includes:
 - [tests/modules/local/extractSharedFastq/main.nf.test](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-682d8a4737a2f00f4e42525869d27dd5a45d9993e8a2d153914782ac23555607)
- [tests/modules/local/maskRead/main.nf.test](tests/modules/local/maskRead/main.nf.test)
- [tests/modules/local/processViralMinimap2Sam/main.nf.test](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-aace11acc6cbbdbc65f26cf6f37bdb20bffd7c4141e488070eb405597732ba3f)
- [tests/subworkflows/local/extractViralReadsONT/main.nf.test](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-65364e7a7125c6e37fca5b286ac976c9311a605cb498613cf1c8b21bf050de80)
 - For this test I added a partly [low-complexity fastq test file](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-65364e7a7125c6e37fca5b286ac976c9311a605cb498613cf1c8b21bf050de80)

- Expanded tests in [tests/modules/local/minimap2/align.nf.test](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-a012351a3206972e7c72cbd0d5550d7ba3c9a60fe54715277baa71341bc53dcc) to test both viral, ribosomal and human index.

Other changes
- Added a new process for minimap2 that does non-streamed alignment. The index for contaminants is very large, which makes minimap2 use "multi-index" alignment, which, to my knowledge, interferes with how minimap2 handles stdout, interfering with streaming.
I played around with this a bunch and gave up, given that this is used only once.
- That process has its own test, here:[tests/modules/local/minimap2/align_non_streamed.nf.test](https://github.com/naobservatory/mgs-workflow/pull/205/files#diff-951bcd5ca7422de008ab955b07a1f22204e3155b1d2382baeff78f923d9e3069)

simonleandergrimm commented on workflows/run_dev_se.nf:
```diff
@@ -23,29 +26,49 @@ nextflow.preview.output = true
     // Count reads in files
     COUNT_TOTAL_READS(samplesheet_ch, params.single_end)
 
+    // Extract viral reads
```
Ah right, I missed that. Fixed!

simonleandergrimm commented on subworkflows/local/extractViralReadsONT/main.nf:
```diff
@@ -0,0 +1,74 @@
+        virus_fastq_ch = virus_minimap2_ch.reads_mapped
+
+        // Pull out clean reads from mapped reads
+        clean_matched_subset_ch = EXTRACT_SHARED_FASTQ_READS(virus_fastq_ch.join(filtered_ch.reads))
```
aha, so you mean we should just pass the full read file into PROCESS_VIRAL_MINIMAP2_SAM? 

Seems right. I originally added this step as `PROCESS_VIRAL_MINIMAP2_SAM` was working on a merged SAM, requiring a merged FASTQ, which would have been huge if we had taken all the raw reads in one file. But on a per-file basis, this is feasible.

willbradshaw replied:
Yes, that's what I mean. Sounds good!

simonleandergrimm replied:
FWIW, I still need this to feed BLAST though.

willbradshaw replied:
If you're masking reads for Minimap2, why don't you want to mask them for BLAST?

simonleandergrimm replied:
I don't think BLAST gets tripped up by non-masked reads as much?

willbradshaw replied:
On reflection, I think this actually isn't very important. We can leave it as is.

simonleandergrimm commented on modules/local/minimap2/main.nf:
```diff
@@ -59,6 +60,58 @@ process MINIMAP2 {
+        #   - First branch (samtools view -u -f 4 -) filters SAM to unaligned reads and saves FASTQ
+        #   - Second branch (samtools view -u -F 4 -) filters SAM to aligned reads and saves FASTQ
+        #   - Third branch (samtools view -h -F 4 -) also filters SAM to aligned reads and saves SAM
+        minimap2 -a ${idx} ${reads} --split-prefix "mm2_split_" > complete_sam.sam
```
Yeah that's how I understand it. Also fed the whole minimap2 code base into Gemini and it gave me the info below.

Looking at the minimap2 manual after seeing the Gemini results, I could increase the total amount of bases that minimap2 indexing loads into memory, which could give us a non multi-part index. But then RAM requirements are also high for aligning? Let me know how important this is to you.

Gemini:

> Minimap2 creates and handles multi-part indices to efficiently index and map against very large target sequence collections that exceed available RAM. Here's a breakdown of how it works:

> 1. Indexing Large Targets

> Batching: When indexing a large target sequence (e.g., a large genome), minimap2 can load only a portion of the target sequences into RAM at a time. This is controlled by the -I option which specifies the maximum number of target bases to load into RAM.
Multi-Part Index: If the total size of the target sequences is larger than the specified batch size, minimap2 creates a multi-part index. Each part of the index corresponds to a different batch of target sequences.
Index Reader: The mm_idx_reader_t structure and related functions (mm_idx_reader_open, mm_idx_reader_read, etc.) are used to manage the reading and building of these multi-part indices. The mm_idx_reader_read function, in particular, is designed to be called repeatedly to process each part of the index.
> 2. Mapping with Multi-Part Indices

> Iterative Mapping: When mapping query sequences against a multi-part index, minimap2 may need to read the query sequences multiple times. This is because each part of the index only contains information about a subset of the target sequences.
Split Prefix: The --split-prefix option is crucial for handling multi-part indices. It allows minimap2 to create temporary files to store intermediate mapping results for each part of the index.
Merging Results: After mapping against each part of the index, the temporary files are merged to produce the final mapping results. The mm_split_merge function is responsible for this merging process.
Rid Shift: When merging results from multi-part indices, it's essential to adjust the reference sequence IDs (rids) because each index part has its own local rid space. The rid_shift array is used to keep track of the necessary adjustments.
In essence, minimap2 handles multi-part indices by:

> Breaking down the indexing and mapping processes into smaller, manageable batches.
Creating temporary files to store intermediate results.
Using an index reader to efficiently process each part of the index.
Merging the results from each part to generate the final output.
This approach allows minimap2 to scale to very large datasets without requiring excessive amounts of RAM, although it may come at the cost of increased processing time due to the need for multiple passes and merging.

willbradshaw replied:
Another option is to separate out the contaminant indices and run against each of them in turn.

I think both of these are probably worth a try, but as long as the current solution works I think we can merge this and improve it later (make an issue?)

I'm currently confused about the format here, I'd expect `> complete_sam.sam` to lead to each split SAM overwriting the previous one rather than concatenating them together.

simonleandergrimm replied:
Added an issue! 

> I'm currently confused about the format here, I'd expect > complete_sam.sam to lead to each split SAM overwriting the previous one rather than concatenating them together.

I don't fully understand how minimap2 is handling stdout here. But it passes tests!

simonleandergrimm commented on modules/local/maskRead/main.nf:
```diff
@@ -30,6 +30,35 @@ process MASK_GENOME_FASTA {
+        par="window=!{window_size} entropy=!{entropy}"
+
+        # Execute with streaming approach
+        zcat -f !{reads} | bbmask.sh in=stdin.fastq out=stdout.fastq ${par} | gzip > ${out}
```
Are you fine with me using bbmask here?

willbradshaw replied:
If it seems to be working well, that's fine.

simonleandergrimm commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,250 @@
+    out["query_sequence_clean"] = query_seq_clean
+    out["query_quality_clean"] = query_qual_clean
+    out["query_length_clean"] = query_len_clean
+
```
Fixed

simonleandergrimm commented on tests/modules/local/maskRead/main.nf.test:
```diff
@@ -0,0 +1,56 @@
+            def masked_seqs = fastq_out.sequences
+            assert masked_seqs.take(5).every { it.contains("N") }
+
+            // The remaining sequences should not be masked
```
Done.

simonleandergrimm commented:
@willbradshaw am I good to start updating the change log? I will first want to run the pipeline on a larger ONT dataset though to see if things are working fine.

willbradshaw commented on modules/local/processViralMinimap2Sam/resources/usr/bin/process_viral_minimap2_sam.py:
```diff
@@ -0,0 +1,250 @@
+    out["minimap2_taxid_primary"] = reference_taxid
+
+    # Marking host-virus status
+    try:
```
@simonleandergrimm the comment thread seems to have gone missing, but I still want to cut this.

simonleandergrimm replied:
Done.

willbradshaw commented:
@simonleandergrimm did another review, the only substantive change I still want to see before provisional approval (conditional on fixing conflicts with dev, tests passing, etc) is removing viral infection status from the SAM script.

simonleandergrimm commented:
Great thanks. Already merged dev. I'm now running all module and subworkflow tests.

simonleandergrimm commented on tests/modules/local/processViralMinimap2Sam/main.nf.test:
```diff
@@ -0,0 +1,75 @@
+            def tab_out = path(process.out.output[0][1]).csv(sep: "\t", decompress: true)
+            assert tab_out.columnCount > 0
+            assert tab_out.rowCount > 0
+            // TODO: Add more detailed tests of file content (requires parsing gzipped SAM file)
```
Done

simonleandergrimm commented:
Alright, I created a CHANGELOG entry. Now running all local tests.

willbradshaw commented on CHANGELOG.md:
```diff
@@ -1,3 +1,19 @@
+# v2.8.2.0-dev
```
What makes this a results bump? I don't see anything that breaks backwards comparability of existing results.

simonleandergrimm replied:
Right, I changed it to v2.8.1.3-dev

willbradshaw replied:
Cool, in that case you can just add this as part of the entry for 2.8.1.2-dev.

simonleandergrimm commented:
Now running all local tests.

willbradshaw commented on tests/subworkflows/local/extractViralReadsONT/main.nf.test:
```diff
@@ -0,0 +1,55 @@
+            assert sam_ids.size() > 0
+            assert sam_ids == hits_hv_read_ids
+
+            // Check the the human and contam filtering work and produce FASTQ files that are subsets of the prior FASTQ file.
```
There aren't any tests in this section

simonleandergrimm replied:
Fixed.

willbradshaw replied:
Great, can you rerun the test and confirm it still passes?

simonleandergrimm replied:
Confirming that it passes.

willbradshaw commented on workflows/run_dev_se.nf:
```diff
@@ -23,30 +26,60 @@ nextflow.preview.output = true
+    } else {
+        EXTRACT_VIRAL_READS_SHORT(samplesheet_ch, params.ref_dir, kraken_db_path, params.bt2_score_threshold, params.adapters, params.host_taxon, "1", "24", "viral", params.bracken_threshold)
+        hv_tsv_ch = EXTRACT_VIRAL_READS_SHORT.out.hits_filtered
+        hv_fastqs = EXTRACT_VIRAL_READS_SHORT.out.hits_fastq
```
Consider publishing this to the intermediates directory

simonleandergrimm replied:
Done.

simonleandergrimm replied:
It does!

simonleandergrimm commented:
I've now run all subworkflow and module tests. They all pass (four failed, one of them was due to outdated samplesheet loading, the other were nf-test dir errors that went away after running again):

```
================================================================================
Test Summary
================================================================================
   ✓ tests/modules/local/addFixedColumn/main.nf.test
   ✓ tests/modules/local/addSampleColumn/main.nf.test
   ✓ tests/modules/local/annotateVirusInfection/main.nf.test
   ✓ tests/modules/local/bbduk/bbduk.nf.test
   ✓ tests/modules/local/bbduk/bbduk_hits.nf.test
   ✓ tests/modules/local/bbmerge/main.nf.test
   ✓ tests/modules/local/blast/main.nf.test
   ✓ tests/modules/local/bowtie2/main.nf.test
   ✓ tests/modules/local/bracken/main.nf.test
   ✓ tests/modules/local/concatenateFiles/main.nf.test
   ✓ tests/modules/local/concatenateTsvs/concatenate_tsvs.nf.test
   ✓ tests/modules/local/concatenateTsvs/concatenate_tsvs_labeled.nf.test
   ✓ tests/modules/local/convertFastqFasta/main.nf.test
   ✓ tests/modules/local/countReads/main.nf.test
   ✓ tests/modules/local/cutadapt/main.nf.test
   ✓ tests/modules/local/extractSharedFastq/main.nf.test
   ✓ tests/modules/local/extractViralHitsToFastq/main.nf.test
   ✓ tests/modules/local/extractViralHitsToFastqNoref/main.nf.test
   ✓ tests/modules/local/fastp/main.nf.test
   ✓ tests/modules/local/filterBlast/main.nf.test
   ✓ tests/modules/local/filterTsv/main.nf.test
   ✓ tests/modules/local/filterVirusReads/main.nf.test
   ✓ tests/modules/local/headTsv/main.nf.test
   ✓ tests/modules/local/interleaveFastq/main.nf.test
   ✓ tests/modules/local/joinFastq/main.nf.test
   ✓ tests/modules/local/joinRiboRef/main.nf.test
   ✓ tests/modules/local/joinTsvs/main.nf.test
   ✓ tests/modules/local/kraken/main.nf.test
   ✓ tests/modules/local/markAlignmentDuplicates/main.nf.test
   ✓ tests/modules/local/maskRead/main.nf.test
   ✓ tests/modules/local/minimap2/align.nf.test
   ✗ tests/modules/local/minimap2/align_non_streamed.nf.test
   ✓ tests/modules/local/minimap2/index.nf.test
   ✓ tests/modules/local/partitionTsv/main.nf.test
   ✓ tests/modules/local/processKrakenViral/main.nf.test
   ✓ tests/modules/local/processViralBowtie2Sam/main.nf.test
   ✓ tests/modules/local/processViralMinimap2Sam/main.nf.test
   ✓ tests/modules/local/reheadTsv/main.nf.test
   ✓ tests/modules/local/replaceString/replace_string.nf.test
   ✓ tests/modules/local/replaceString/replace_string_compressed.nf.test
   ✓ tests/modules/local/sortFastq/main.nf.test
   ✓ tests/modules/local/sortFile/main.nf.test
   ✓ tests/modules/local/sortTsv/main.nf.test
   ✓ tests/modules/local/subsetFastq/main.nf.test
   ✓ tests/modules/local/summarizeBBMerge/main.nf.test
   ✓ tests/modules/local/summarizeMultiqc/main.nf.test
   ✓ tests/modules/local/unleaveFastq/main.nf.test
   ✓ tests/subworkflows/local/blastViral/main.nf.test
   ✓ tests/subworkflows/local/countTotalReads/main.nf.test
   ✓ tests/subworkflows/local/extractViralReadsONT/main.nf.test
   ✓ tests/subworkflows/local/extractViralReadsShort/main.nf.test
   ✓ tests/subworkflows/local/loadDownstreamData/main.nf.test
   ✓ tests/subworkflows/local/loadSampleSheet/main.nf.test
   ✓ tests/subworkflows/local/makeVirusTaxonomyDB/main.nf.test
   ✗ tests/subworkflows/local/markViralDuplicates/main.nf.test
   ✗ tests/subworkflows/local/prepareGroupTsvs/main.nf.test
   ✗ tests/subworkflows/local/profile/main.nf.test
   ✓ tests/subworkflows/local/qc/main.nf.test
   ✓ tests/subworkflows/local/runQc/main.nf.test
   ✓ tests/subworkflows/local/subsetTrim/main.nf.test
   ✓ tests/subworkflows/local/taxonomy/main.nf.test

----------------------------------------
Total:  61 tests
Passed: 57 tests
Failed: 4 tests
----------------------------------------
```

Re-running failed tests:
```
[ec2-user@ip-172-31-40-150 mgs-workflow]$ nf-test test tests/subworkflows/local/extractViralReadsONT/main.nf.test  --verbose

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Load .nf-test/plugins/nft-csv/0.1.0/nft-csv-0.1.0.jar
Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar
Load .nf-test/plugins/nft-fasta/1.0.0/nft-fasta-1.0.0.jar
Load .nf-test/plugins/nft-bam/0.5.0/nft-bam-0.5.0.jar

Test subworkflow EXTRACT_VIRAL_READS_ONT

  Test [a61e0412] 'Should run without failures' 
SUCCESS: Executed 1 tests in 103.92s

[ec2-user@ip-172-31-40-150 mgs-workflow]$ nf-test test tests/subworkflows/local/prepareGroupTsvs/main.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Load .nf-test/plugins/nft-csv/0.1.0/nft-csv-0.1.0.jar
Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar
Load .nf-test/plugins/nft-fasta/1.0.0/nft-fasta-1.0.0.jar
Load .nf-test/plugins/nft-bam/0.5.0/nft-bam-0.5.0.jar

Test workflow PREPARE_GROUP_TSVS

  Test [3ccac309] 'Should run without failures' PASSED (41.424s)


SUCCESS: Executed 1 tests in 41.431s

[ec2-user@ip-172-31-40-150 mgs-workflow]$ nf-test test tests/subworkflows/local/markViralDuplicates/main.nf.test --verbose

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Load .nf-test/plugins/nft-csv/0.1.0/nft-csv-0.1.0.jar
Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar
Load .nf-test/plugins/nft-fasta/1.0.0/nft-fasta-1.0.0.jar
Load .nf-test/plugins/nft-bam/0.5.0/nft-bam-0.5.0.jar

Test workflow MARK_VIRAL_DUPLICATES

  Test [87d0aec4] 'Should run without failures (deviation = 1)'
  Test [d1b4b04c] 'Should run without failures (deviation = 2)'
  Test [7c0f737a] 'Should run without failures (deviation = 0)'

SUCCESS: Executed 3 tests in 109.566s
[ec2-user@ip-172-31-40-150 mgs-workflow]$ nf-test test    tests/subworkflows/local/profile/main.nf.test

🚀 nf-test 0.9.2
https://www.nf-test.com
(c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr

Load .nf-test/plugins/nft-csv/0.1.0/nft-csv-0.1.0.jar
Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar
Load .nf-test/plugins/nft-fasta/1.0.0/nft-fasta-1.0.0.jar
Load .nf-test/plugins/nft-bam/0.5.0/nft-bam-0.5.0.jar

Test subworkflow PROFILE

  Test [49a210bc] 'Should run without failures on paired (interleaved) input' PASSED (96.884s)
  Test [a18d33d6] 'Should run without failures on unpaired input' PASSED (84.302s)
```

SUCCESS: Executed 2 tests in 181.194s

simonleandergrimm commented:
Alright, the CHANGELOG is updated. @willbradshaw I believe this is ready to go in.
